### PR TITLE
Convert 1X cookies to use 2X format. Move to root domain.

### PR DESCRIPTION
**Pull request by schwers**
👓  @nramadas @phil303

### CODE REVIEW

File: src/server/session.es6.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+export const COOKIE_OPTIONS = {
+  secure: config.https,
+  secureProxy: config.httpsProxy,
+  httpOnly: true,
+  maxAge: 1000 * 60 * 60 * 24 * 365,
+};
+
+export const setSessionCookies = (ctx, session) => {
+  const payloadBase64 = btoa(JSON.stringify(session));
+  const token = `${payloadBase64}.2`; // version 2
```
Comment (schwers) [inline src/server/session.es6.js]: version constants

File: assets/js/clientLib.es6.js [hunk 2]
```diff
@@ -118,31 +118,26 @@ export function refreshToken (app) {
...
-        Object.assign(app.getState('ctx'), {
-          token: token.token,
-          tokenExpires: token.tokenExpires,
+        Object.assign(app.getState('ctx').redditSession, {
+          accessToken: session.accessToken,
+          expires: session.expires,
         });
 
+        if (session.refreshToken) {
+          Object.asseign(app.getState('ctx').redditSession, {
```
Comment (schwers) [inline assets/js/clientLib.es6.js]: `Object.assign`

File: assets/js/client.es6.js [hunk 3]
```diff
@@ -78,7 +78,7 @@ function initialize(bindLinks) {
   const $body = document.body || document.getElementsByTagName('body')[0];
   setAppMethods(app, $body, render, history);
 
-  if (app.getState('token')) {
+  if (app.getState('session')) {
```
Comment (nramadas) [inline assets/js/client.es6.js]: 🔕  just want to make sure `session` is intentional, and it's not supposed to be `redditSession`.

File: src/server/oauth.es6.js [hunk 4]
```diff
@@ -499,7 +473,8 @@ const oauthRoutes = function(app) {
           resolve();
         }
 
-        const message = obj.message ? `&message=${obj.message}` : '';
+        const message = obj.message ?
```
Comment (nramadas) [inline src/server/oauth.es6.js]: 🔕  why the change?

File: src/server/session.es6.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  if (tokenCookie.indexOf('.') === -1) {
+    if (!oldExpiresCookie || !oldRefreshCookie) {
+      // we have a token cookie that looks like the new format, but we
+      // don't have the expiration and refreshToken, just clear auth cookies
+      clearSessionCookies(ctx.cookies);
+      return;
+    }
+
+    // clear the old tokenExpires and refreshToken cookies
+    clearSessionCookies(ctx.cookies);
```
Comment (nramadas) [inline src/server/session.es6.js]: up to you, but an alternate organization of this code could look like:

```es6
if (tokenCookie.indexOf('.') === -1) {
  // this is an old style token cookie. we want to go ahead and clear all the pertinent
  // session cookies.
  clearSessionCookies(ctx.cookies);

  // if we have both an 'expires' cookie and a 'refresh' cookie, create a new session out of it.
  if (oldExpiresCookie && oldRefreshCookie) {
    const newSession = {
      accessToken: tokenCookie,
      expires: oldExpiresCookie,
      refreshToken: oldRefreshCookie,
    };

    setSessionCookies(ctx, newSession);
    return newSession;
  }
}

```

File: src/server/session.es6.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+};
+
+
+export const SESSION_COOKIES = [
+  'token',
+  'tokenExpires',
+  'refreshToken',
+];
+
+export const clearSessionCookies = cookies => {
```
Comment (nramadas) [inline src/server/session.es6.js]: :no_bell: Nit, but I think it would be nice for `clearSessionCookies` to take in `ctx`, so that its signature is complementary to `setSessionCookies`.

File: src/server/session.es6.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    if (!session || !session.accessToken || !session.refreshToken || !session.expires) {
+      clearSessionCookies(ctx.cookies);
+      return;
+    }
+
+    return session;
+  } catch (e) {
+    // clear the malformed cookies
+    clearSessionCookies(ctx.cookies);
+    // implicit return
```
Comment (nramadas) [inline src/server/session.es6.js]: 🔕  It might be reasonable to just `return` here instead of leaving a comment explaining that the return is implicit :p

File: src/server/session.es6.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+  }
+};
+
+export const COOKIE_OPTIONS = {
+  secure: config.https,
+  secureProxy: config.httpsProxy,
+  httpOnly: true,
+  maxAge: 1000 * 60 * 60 * 24 * 365,
```
Comment (nramadas) [inline src/server/session.es6.js]: 1x code, i know, but should this number live in a constant?

File: src/views/pages/login.jsx [hunk 9]
```diff
@@ -263,15 +263,19 @@ class LoginPage extends BasePage {
 
       // do some redirection here.
       if (res && res.body) {
-        const { token } = res.body.token;
+        const { redditSession } = res.body;
+
+        const ctx = app.getState('ctx');
```
Comment (nramadas) [inline src/views/pages/login.jsx]: 🔕  so, in 1X, state that comes out of `getState` isn't a new reference, right? so, to keep things consistent with how it exists elsewhere, this could read:

```es6
Object.assign(app.getState('ctx'), {
  ...ctx,
  redditSession: {
    ...ctx.redditSession,
    ...redditSession,
  },
});
```

**Comment by nramadas** [comment]
:nail_care:

File: src/server/session.es6.js (hunk 7 above)
Comment (schwers) [inline src/server/session.es6.js]: I think I may have been overzealous with comment details in this, do you have any objection to just removing the comment and not having an explicit return? When I look at it now it doesn't seem necessary

File: src/server/session.es6.js (hunk 8 above)
Comment (schwers) [inline src/server/session.es6.js]: Sure, that's more readable

File: src/server/session.es6.js (hunk 7 above)
Comment (nramadas) [inline src/server/session.es6.js]: nope, that's fine by me too

File: src/server/session.es6.js (hunk 6 above)
Comment (schwers) [inline src/server/session.es6.js]: Having the parameter be `ctx` would be some satisfying symmetry. I liked only requiring `cookies` so callers understand `clearSessionCookies` only needs to use `ctx.cookies`. With that in mind, would you still prefer it taking `ctx`?

File: src/server/oauth.es6.js (hunk 4 above)
Comment (schwers) [inline src/server/oauth.es6.js]: bad-edit, thanks

File: src/server/session.es6.js (hunk 6 above)
Comment (nramadas) [inline src/server/session.es6.js]: i think i would still prefer `ctx`. The other reason being, we understand `ctx` to be something we mutate on a regular basis (and has a clear meaning on our codebase: it's koa ctx). However, `cookies` isn't quite as clear cut. And, in this particular case, though we are calling methods on `cookies`, ultimately, it's to mutate the koa `ctx` object, so passing in `ctx` feel appropriate.

File: assets/js/client.es6.js (hunk 3 above)
Comment (schwers) [inline assets/js/client.es6.js]: good catch

File: src/views/pages/login.jsx (hunk 9 above)
Comment (schwers) [inline src/views/pages/login.jsx]: Yes, it's not technically a new a reference, I dislike that assumption in 1X code so I'll leave it as is. (From a quick look `setState` actually seems more consistent)

Comment (nramadas) [inline src/views/pages/login.jsx]: I'm fine with that 👍 . I was just curious for my own understanding

**Comment by schwers** [comment]
💇

**Comment by nramadas** [comment]
one question, but other than that, LGTM :fish:
